### DISCUSSION TITLE
I converted the tutorial to Jupyter Notebooks 😄 (http://sqla-tutorials-nb.rtfd.io)

### DISCUSSION THREAD

**Discussion by chrisjsewell** [sqlalchemy/sqlalchemy#7138] [Ideas]
Heya, I've found the https://docs.sqlalchemy.org/en/14/tutorial/  super useful, but a few time now I have come back to it to remember / test out certain API features, and it's been annoying to re-load all the requisite code into a python/ipython session.

So, being a maintainer of https://jupyterbook.org, I thought I'd convert it to "executable" content 😅

With https://sqla-tutorials-nb.readthedocs.io, each page is a self-contained notebook (executed on docs build), which you can download or even launch interactively from the site with [Binder](https://mybinder.org/) or [Thebe](https://thebe.readthedocs.io)

Hopefully it is off use to others, and happy to transfer the repo anywhere if necessary

**Comment by CaselIT** [comment]
Hi,

This looks interesting, but having never used it I have some questions:
- does this works directly from the documentation rst or does it require some translation to happen?
- if it uses a generation step is this fully automatic? (or can it be made so with a reasonable effort)
- does the generated code need a dedicated git repository?

Thanks for the interest!

**Reply by chrisjsewell** [reply]
Heya,

As partially mentioned in the https://github.com/chrisjsewell/sqla-tutorials-nb/blob/main/README.md (I should update that),
it still uses sphinx (see https://github.com/chrisjsewell/sqla-tutorials-nb/blob/main/docs/conf.py), but the translation is "semi-automatic":

1. https://github.com/executablebooks/rst-to-myst does the bulk of the work, converting from RST to MyST Markdown
1. You then add "top-matter" to each file, to allow jupytext to understand how to convert them dynamically to Notebooks (see https://jupytext.readthedocs.io/en/latest/formats.html#myst-markdown) 
1. I replaced all the API references to use "conventional" intersphinx references, e.g. `` :func:`_sql.select` `` -> `` {func}`~sqlalchemy.future.select` ``
1. I replaced the `container` directives to use `div` directives (very similar but `container` are not compatible with Bootstrap themes like sphinx-book-theme)
1. Then the main thing was to convert all the `.. sourcecode:: pycon+sql`, to notebook cells.
  For example:
    ```restructuredtext
    .. sourcecode:: pycon+sql
    
        >>> stmt = (
        ...   select(User.fullname).
        ...   join(User.addresses.and_(Address.email_address == 'pearl.krabs@gmail.com'))
        ... )
        >>> session.execute(stmt).all()
        {opensql}SELECT user_account.fullname
        FROM user_account
        JOIN address ON user_account.id = address.user_id AND address.email_address = ?
        [...] ('pearl.krabs@gmail.com',){stop}
        [('Pearl Krabs',)]
    ```
    
    goes to
    
    ````markdown
    ```{code-cell} ipython3
    stmt = (
      select(User.fullname).
      join(User.addresses.and_(Address.email_address == 'pearl.krabs@gmail.com'))
    )
    session.execute(stmt).all()
    ```
    ````
    
    (then the notebook execution during sphinx-build handles adding the actual outputs)
    
1. Finally, for each document, they mostly required adding an extra initial code cell, to setup the initial database, i.e. in the sqlalchemy docs each tutorial page assumes that you have already run all the code from the previous pages, whereas in mine each page is "self-contained"

1. Oh and a few extra "flourishes" using https://sphinx-design.readthedocs.io/en/furo-theme/

So yeh, in theory it would be possible to have this all in the source sqlalchemy docs, but that would not be completely straightforward 😬

**Reply by CaselIT** [reply]
Thanks for the rundown! Looking at it, points 1,2,3,4 seem to be automatize, 6 is a one time thing, 5 seems also doable with a bit more effort.

For my part I think we could add a link to this notebook friendly version form the tutorial docs. Then if we have community help we could look into automatic the migration steps to integrate them in sqlalchemy doc releases.

@zzzeek opinions?

**Comment by zzzeek** [comment]
wow this is beautiful, i havent done the juypiter part of it but just you did all the styling and all that, you emulated the CSS we had and all that to accomplish this?   i mean this looks really refreshing, with all the modern stuff like "copy text" controls and all that.

we're also doing a rework of our own docs but so far we're just trying to modernize the JS and layout to use more modern tooling, without yet changing layout too much.    im still pretty attached (reasonably or not) to the general idea of our current design but being able to opt into some more modern controls and maybe trying to "brighten" the design a bit (im not sure if our left hand nav is very usable, the split left/right nav here might be better) seems appealing.

**Reply by chrisjsewell** [reply]
Thanks!

For the core/orm boxes, I just copied your existing CSS 😉 https://github.com/chrisjsewell/sqla-tutorials-nb/blob/main/docs/_static/custom.css

But yeh lots going on in the wonderful world of sphinx HTML themes lol, particularly with things like https://sphinx-book-theme.readthedocs.io, https://pradyunsg.me/furo/, https://pydata-sphinx-theme.readthedocs.io, and https://sphinx-design.readthedocs.io
(we are all very much working together on this: https://github.com/executablebooks/meta/issues/279)

Feel free to ping me with any questions, or if you need any assistance (time permitting 😬), since it's always nice to give back to great tools that I'm using 😄

**Comment by Winson-Huang** [comment]
wonderful!
